src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringInterpolation.cs

```diff
@@ -38,38 +39,67 @@ private BoundExpression RewriteInterpolatedStringConversion(BoundConversion conv
+            // var handler = new HandlerType(baseStringLength, numFormatHoles, ...InterpolatedStringHandlerArgumentAttribute parameters, <optional> out bool handlerIsValid);
+            var construction = (BoundObjectCreationExpression)data.Construction;
+
+            BoundLocal? appendShouldProceedLocal = null;
+            bool ownsAppendShouldProceedLocal = false;
+            if (data.HasTrailingHandlerValidityParameter)
+            {
+                Debug.Assert(construction.ArgumentRefKindsOpt[^1] == RefKind.Out);
+
+                BoundInterpolatedStringArgumentPlaceholder trailingParameter = data.ArgumentPlaceholders[^1];
+                if (!HasPlaceholderReplacement(trailingParameter))
+                {
+                    ownsAppendShouldProceedLocal = true;
+                    TypeSymbol localType = trailingParameter.Type!;
+                    Debug.Assert(localType.SpecialType == SpecialType.System_Boolean);
+                    var outLocal = _factory.SynthesizedLocal(localType);
+                    appendShouldProceedLocal = _factory.Local(outLocal);
+
+                    AddPlaceholderReplacement(trailingParameter, appendShouldProceedLocal);
+                }
+                else
+                {
+                    // We're already processing this node, so use the local VisitArguments will have allocated for us
+                    appendShouldProceedLocal = (BoundLocal)PlaceholderReplacement(trailingParameter);
+                    Debug.Assert(appendShouldProceedLocal.Type.SpecialType == SpecialType.System_Boolean);
+                }
+            }
+
+            var handlerConstructionAssignment = _factory.AssignmentExpression(builderTemp, (BoundExpression)VisitObjectCreationExpression(construction));
 
             var usesBoolReturn = data.UsesBoolReturns;
-            var builderPatternExpressions = ArrayBuilder<BoundExpression>.GetInstance(usesBoolReturn ? 2 : node.Parts.Length);
-            builderPatternExpressions.Add(builderConstruction);
-            BoundExpression? appendBinaryExpression = null;
+            var resultExpressions = ArrayBuilder<BoundExpression>.GetInstance(node.Parts.Length + 1);
             foreach (var currentPart in node.Parts)
             {
                 var appendCall = (BoundCall)currentPart;
                 Debug.Assert(usesBoolReturn == (appendCall.Method.ReturnType.SpecialType == SpecialType.System_Boolean));
 
-                var rewrittenArgs = VisitList(appendCall.Arguments);
-                var rewrittenAppendCall = MakeCall(
+                // The append call itself could have and interpolated string conversion that uses the builder local as the receiver
+                // passed to a nested construction call. This needs to affect the current call to ensure side effects are
+                // preserved, but shouldn't affect any subsequent calls.
+                BoundExpression appendReceiver = builderTemp;
```

Comment:
>appendReceiver [](http://example.com/codeflow?start=32&length=14)

It doesn't look like we are mutating the value or passing the local ByRef. What exact purpose does this local have?

Reply:
Will be used in the next revision.